### ISSUE TITLE
StreamField ValidationError not shown at specific field or indicated in content tab

### ISSUE THREAD

**Issue by lb-** [wagtail/wagtail#4122]
### Issue Summary

When validating ``StreamField`` using the ``validators`` attribute, validation errors are indicated to the user but no field specific errors/text is shown.

This may be a related to this issue: #3441
Found this out after attempting to answer this Stack Overflow question: https://stackoverflow.com/questions/47760595/validate-no-duplicates-in-streamfield

### Steps to Reproduce

1. Start on the BakeryDemo project
2. Edit the base/models.py file - ``StandardPage`` model 
3. Create a validator as per [Django docs](https://docs.djangoproject.com/en/2.0/ref/forms/validation/#using-validators)
```
def validate_headings_are_unique(value):
    items = [data[1] for data in value.stream_data if 'heading' in data[0]]
    if len(set(items)) != len(items):
        raise ValidationError(
            'All headings must have unique values: %(value)s',
            code='invalid',
            params={'value': items},
        )
```
4. Edit the field body as such (using the validators attribute):
```
    body = StreamField([
        ('heading2', CharBlock(label='h2')),
        ('heading3', CharBlock(label='h3')),
        ('paragraph', RichTextBlock())
    ], blank=True, validators=[validate_headings_are_unique])
```
5. Edit a page using the StandardPage model
6. Enter two headings into the Body field, a H2 with value 'foo' and a H3 with value 'foo'
6a. Expected: Error treated like other validation errors on non StreamField fields...
   * Error at top: The page could not be saved due to validation errors
   * Error indicator on the content tab (1 in red circle)
   * Error on the field with the error text shown
6b. Actual: Error at the top of the page only, with no indication of what 'field' this error occurred on.
![wagtail - editing standard page - editor error](https://user-images.githubusercontent.com/1396140/33927178-5cf46644-e035-11e7-8e6a-dbcf2974e550.png)
Note: Page error only, no error on the content tab and no error at the 'Body' panel

### Technical details

* Python version: 3.4.3
* Django version: 1.11.6
* Wagtail version: 2.0a0

### Potential Solution 1 - Minor fix + Documentation of ``StreamBlockValidationError``

After digging around a bit, I can see that validation errors can be made with the ``StreamBlockValidationError`` from here:
https://github.com/wagtail/wagtail/blob/db6d36845f3f2c5d7009a22421c2efab9968aa24/wagtail/core/blocks/stream_block.py#L21-L29

The revised validator looks like:

```
def validate_headings_are_unique(value):
    items = [data[1] for data in value.stream_data if 'heading' in data[0]]
    if len(set(items)) != len(items):
        raise StreamBlockValidationError(
            non_block_errors=ValidationError(
                'All headings must have unique values: %(value)s',
                code='invalid',
                params={'value': items},
            ),
        )
```

This mostly works, the error appears at the top of the Streamfield, before all blocks. However the content tab does not have an indicator of the error being in that tab.

So it would be hard for the user to find this error if it was not on the primary tab. There is another issue here, the error div appears to be in the wrong spot, it appears under the (+) icon, which is odd, it should probably appear above it so that adding new items to the start of the StreamField will not appear above the errors.

![wagtail - editing standard page - non_block_errors location](https://user-images.githubusercontent.com/1396140/33928193-2547e0f0-e039-11e7-8419-fe63e9cb68dd.png)

This would be good if it worked well and was documented, as it looks like you could also dictate where you want the error to appear, but I could not work out how to do that.

### Potential Solution 2 - handle ValidationErrors better

Ideally we would do the above solution, and handle the normal case of ValidationError as if it was a `StreamBlockValidationError` with `non_block_errors` set to whatever errors are passed in.

**Comment by phildexter** [comment]
This should be fixed by the work we're doing this year on the page editor - new designs and discussions about them can be seen [here](https://github.com/wagtail/wagtail/discussions/7739).

**Comment by gasman** [comment]
Have confirmed that this now works on current main, with the validator function updated as follows (to account for `stream_data` being deprecated since 2.12):

```python
def validate_headings_are_unique(value):
    items = [
        block.value['heading_text'] for block in value
        if block.block_type == 'heading_block'
    ]
    if len(set(items)) != len(items):
        raise ValidationError(
            'All headings must have unique values: %(value)s',
            code='invalid',
            params={'value': items},
        )
```

This would presumably have been fixed by #10131, as this made it possible for the top-level StreamBlock to accept plain ValidationErrors rather than requiring a StreamBlockValidationError.
